### PR TITLE
hal: nuvoton: add nuvoton m460 series hal library

### PR DESCRIPTION
**Pull request by cyliangtw**
This PR is to add Nuvoton M460 SoC series hal library. Later, Nuvoton MCU product line software team will push the corresponding PR of Zephyr-rtos for M460 SoC series.
About M460, it's ARM® Cortex®-M4F core running up to 200 MHz with SRAM 512KB & ROM 1024KB.

In this PR, except extracting Stdriver and Device support from M460 BSP, it also create pin-mux header file for Zephyr pinctrl.

Details of Nuvoton M460 BSP as below:
URL: https://github.com/OpenNuvoton/M460BSP
Version: v3.00.001 (SHA: a8a785ccf932485b644f8482ba136051c8579caa)
Purpose: Add support for Nuvoton M46x SoCs
Maintained-by: Nuvoton MCU product line

Signed-off-by:  cyliangtw [cyliang@nuvoton.com](mailto:cyliang@nuvoton.com)

@carlescufi , @ssekar15, please help review

### CODE REVIEW

**Review by ssekar15 — COMMENTED** [review]
Please add details about original repo and release version from SDK in commit message.

**Review by ssekar15 — COMMENTED** [review]
Please sync SDK based on release, which help to avoid conflict of partial update of Stddriver and reset out of sync. I would suggest to drop this change set.

File: m46x/StdDriver/src/fmc.c
Comment (ssekar15) [inline m46x/StdDriver/src/fmc.c]: why it is not part of StdDriver commit !

File: m46x/StdDriver/drv_emac/synopGMAC_network_interface.h
Comment (ssekar15) [inline m46x/StdDriver/drv_emac/synopGMAC_network_interface.h:4]: Apache-2.0 license  please

File: m46x/StdDriver/drv_emac/synopGMAC_plat.h
Comment (ssekar15) [inline m46x/StdDriver/drv_emac/synopGMAC_plat.h:4]: Apache-2.0 license please

File: m46x/StdDriver/drv_emac/synopGMAC_network_interface.c
Comment (ssekar15) [inline m46x/StdDriver/drv_emac/synopGMAC_network_interface.c:4]: Please mention Apache-2.0 license as below in the each file  header comments

SPDX-License-Identifier: Apache-2.0

File: m46x/StdDriver/drv_emac/synopGMAC_Dev.c
Comment (ssekar15) [inline m46x/StdDriver/drv_emac/synopGMAC_Dev.c:4]: Apache-2.0 license please

File: m46x/StdDriver/src/fmc.c
Comment (cyliangtw) [reply m46x/StdDriver/src/fmc.c]:
```
$git log ./m46x/StdDriver/src/fmc.c
Author: cyliangtw <cyliang@nuvoton.com>
Date:   Fri Dec 23 17:37:04 2022 +0800
    Sync M467 STD drivers 2022/12/22
commit 6461402934142dd8ef2b7c91c9a69405f851adff
Author: cyliangtw <cyliang@nuvoton.com>
Date:   Tue Sep 13 19:34:35 2022 +0800
    NuMaker M46x support flash controller FMC driver
```
In the previous commit: 
I tried to follow our Nuvoton FMC API interface
 of other series [M480 FMC API](https://github.com/OpenNuvoton/M480BSP/blob/master/Library/StdDriver/src/fmc.c#L259), so I modified it at that time.

However, after dicussion with our M460 SDK owner, not change it in near future, so my last commit is fullly same with M460 StdDriver on 2022/12/23.

Comment (cyliangtw) [reply m46x/StdDriver/src/fmc.c]: Rebase & fixed in commit a99da0a4ec3cef717e5c3c2b5dbfa7747fa7ddb4

File: m46x/StdDriver/drv_emac/synopGMAC_network_interface.h
Comment (cyliangtw) [reply m46x/StdDriver/drv_emac/synopGMAC_network_interface.h:4]: This code was developed by Synopsys to be tested with Linux OS (Kernel Version 2.6.xy). And Synopsys BU confirmed it's fine to add extra license string as:
```
/* ===================================================================================
 * Copyright (c) <2009> Synopsys, Inc.
 * SPDX-License-Identifier: Apache-2.0
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of

```


Fixed in commit 17e131d6cf28744a149250e72113611fbe278370

**Comment by cyliangtw** [comment]
> Please sync SDK based on release, which help to avoid conflict of partial update of Stddriver and reset out of sync. I would suggest to drop this change set.

Rebase & fixed in commit d1b4db369496af3b9279b45fd2ce9fbe18596776

File: m46x/StdDriver/drv_emac/synopGMAC_plat.h
Comment (cyliangtw) [reply m46x/StdDriver/drv_emac/synopGMAC_plat.h:4]: Fixed in commit https://github.com/zephyrproject-rtos/hal_nuvoton/commit/17e131d6cf28744a149250e72113611fbe278370

File: m46x/StdDriver/drv_emac/synopGMAC_network_interface.c
Comment (cyliangtw) [reply m46x/StdDriver/drv_emac/synopGMAC_network_interface.c:4]: Fixed in commit https://github.com/zephyrproject-rtos/hal_nuvoton/commit/17e131d6cf28744a149250e72113611fbe278370

File: m46x/StdDriver/drv_emac/synopGMAC_Dev.c
Comment (cyliangtw) [reply m46x/StdDriver/drv_emac/synopGMAC_Dev.c:4]: Thanks of your kindly review.
Fixed in commit https://github.com/zephyrproject-rtos/hal_nuvoton/commit/17e131d6cf28744a149250e72113611fbe278370

**Comment by cyliangtw** [comment]
> Please add details about original repo and release version from SDK in commit message.

Thanks of your review and advice.

Rebase & added SDK URL & version in commit d1b4db369496af3b9279b45fd2ce9fbe18596776

**Comment by cyliangtw** [comment]
@ssekar15 , I revised the commit FMC driver [a99da0a](https://github.com/zephyrproject-rtos/hal_nuvoton/pull/5/commits/a99da0a4ec3cef717e5c3c2b5dbfa7747fa7ddb4) to resolve un-sync SDK issue. More, also followed your advice to add details of SDK version/url on commit message and add Apache-2.0 license on EMAC driver.

Please help review or give me some comment.

File: m46x/StdDriver/CMakeLists.txt
Comment (ssekar15) [inline m46x/StdDriver/CMakeLists.txt:7]: please remove below lines part of this commit, be it handled by other separate commits

zephyr_library_sources(src/clk.c)
zephyr_library_sources(src/gpio.c)
zephyr_library_sources(src/uart.c)
#zephyr_library_sources_ifdef(CONFIG_HAS_NUMAKER_UART src/uart.c)

**Review by ssekar15 — COMMENTED** [review]
Thanks for cleanup, please consider below few comments
1. please follow commit header format 
    hal: NuMaker M46x:
    (or)
    driver: emac:  
2. add origin repo, version and sha details in commit "Add Nuvoton M460 series hal lib including STD drivers"
3. squash "Add Apache-2.0 license in emac files"  to "NuMaker M46x support EMAC controller driver"

File: m46x/StdDriver/CMakeLists.txt
Comment (cyliangtw) [reply m46x/StdDriver/CMakeLists.txt:7]: It's done to remove it from this commit and put on the commit  f2302097492 `add uart/gpio controller`.

**Comment by cyliangtw** [comment]
@ssekar15 , thanks of your 2nd review. It's great to unify the header format and I also refer Zephyr style to revise the commit message.

**Comment by ssekar15** [comment]
Looks fine for me

**Review by ssekar15 — APPROVED** [review]

**Comment by cyliangtw** [comment]
@carlescufi , please help give me some comment.

**Comment by cyliangtw** [comment]
@ssekar15 , @carlescufi 
Hi Carlescufi, if there is no comment on this PR, could you help to merge it ?

File: CMakeLists.txt
Comment (stephanosio) [inline CMakeLists.txt]: Please add a new line at the end of file.

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (cyliangtw) [reply CMakeLists.txt]: @stephanosio , thanks of your advice.
It's fixed in commit [d57098a](https://github.com/zephyrproject-rtos/hal_nuvoton/pull/5/commits/d57098abaf62c532f33ea1d8cd3c6f144ba0a046).

Comment (stephanosio) [reply CMakeLists.txt]: > @stephanosio , thanks of your advice.
> It's fixed in commit [d57098a](https://github.com/zephyrproject-rtos/hal_nuvoton/pull/5/commits/d57098abaf62c532f33ea1d8cd3c6f144ba0a046).

Thanks. Please squash that commit into `hal: NuMaker M46x: support external dts or *.h of module hal-M46x` since that change was introduced in this very PR.

Comment (cyliangtw) [reply CMakeLists.txt]: I see and squash this commit into `hal: NuMaker M46x: support external dts or *.h of module hal-M46x` [77ebdca](https://github.com/zephyrproject-rtos/hal_nuvoton/pull/5/commits/77ebdca7b9ca2a2f5d36dc4356eb6c220c5fe0d7)

**Review by stephanosio — APPROVED** [review]

**Comment by cyliangtw** [comment]
@stephanosio, thanks of your review and approval.
Next, who could help to merge this PR ?

**Comment by stephanosio** [comment]
Will merge this after the merge window opens for v3.4 in a few weeks. Please ping me again if I forget to merge it.

**Comment by cyliangtw** [comment]
@stephanosio,  mainline version updated as 3.3.99 [VERSION: Bump to 3.3.99](https://github.com/zephyrproject-rtos/zephyr/commit/5cb53ee1e96224c0b4f802dc7e3026611471b754)
 Is it the time "merge window opens for v3.4" ?

**Comment by stephanosio** [comment]
> @stephanosio, mainline version updated as 3.3.99 [VERSION: Bump to 3.3.99](https://github.com/zephyrproject-rtos/zephyr/commit/5cb53ee1e96224c0b4f802dc7e3026611471b754) Is it the time "merge window opens for v3.4" ?

Correct. Merge window is now open for 3.4.

However, I see that there is currently no Zephyr side PR that pulls this change (https://github.com/zephyrproject-rtos/zephyr/pull/53471 was closed?).

There needs to be a Zephyr side PR that pulls this change and passes the CI for this to be merged.

**Comment by cyliangtw** [comment]
> There needs to be a Zephyr side PR that pulls this change and passes the CI for this to be merged.

There are too many commits in https://github.com/zephyrproject-rtos/zephyr/pull/53471, it not fulfill the typical style of zephyr side PR.
OK, I see and I will split small one.

**Comment by cyliangtw** [comment]
@stephanosio , the relative Szephyr side PR is [zephyr/pull/55406](https://github.com/zephyrproject-rtos/zephyr/pull/55406). 
My local environment could pass the CI tests.
After you merge it, I will change \zephyrproject-rtos\zephyr\west.yaml .

**Comment by cyliangtw** [comment]
@stephanosio ,  Zephyr side PR https://github.com/zephyrproject-rtos/zephyr/pull/55406 that pulls this change and it could pass the CI on my local environment.

Could you help to merge this hal PR ?

**Comment by stephanosio** [comment]
The PR needs to pass CI (i.e. not just your local build attempt).

**Comment by cyliangtw** [comment]
@stephanosio,

In Zephyr side PR [#55406](https://github.com/zephyrproject-rtos/zephyr/pull/55406), it require this hal PR's `/hal_nuvoton/m46x/`. 
So global CI test prompt "No such file" error as:
`/__w/zephyr/zephyr/dts/arm/nuvoton/m46x.dtsi:11:10: fatal error: m46x/sys.h: No such file or directory
   11 | #include "m46x/sys.h"
      |          ^~~~~~~~~~~~`


According to snippet of zephyrproject-rtos/zephyr/west.yml, not include m46x yet:
```
    - name: hal_nuvoton
      revision: 0a1f153c433f5f637a4490651bdda6d966de3b99
      path: modules/hal/nuvoton

```
As result, environmen set up of global twister-build will not include hal_nuvoton/m46x:
`Cloning into '/__w/zephyr/modules/hal/nuvoton'...
done.
HEAD is now at 0a1f153 hal: nuvoton: add pinctrl dtsi file for numicro m480 series`

Thus, [#55406](https://github.com/zephyrproject-rtos/zephyr/pull/55406) CI test must rely on this hal PR and west.yml .
Before merge this hal PR, is there any way to include hal_nuvoton/m46x in CI environmen set up ?

**Comment by stephanosio** [comment]
> Before merge this hal PR, is there any way to include hal_nuvoton/m46x in CI environmen set up ?

Please read https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules.

**Comment by cyliangtw** [comment]
@stephanosio , Thanks of your advice.

CI detected some un-int var. warnings in Nuvoton M460 BSP and I pushed owner to fix it.
So, I sync Nuvoton M460 BSP and add one more commit to resolve  https://github.com/zephyrproject-rtos/zephyr/pull/55406 CI warning/errors.
After Zephyr side PR could pass "Run tests with twister / Publish Unit Tests Results", I rebase    it.

image

**Review by stephanosio — APPROVED** [review]

**Pull request merged**
